demo/app/demo-cusom-items/index.js

```diff
@@ -0,0 +1,234 @@
+    } else if (visibleTimeEnd > maxTime) {
+      updateScrollCanvas(maxTime - (visibleTimeEnd - visibleTimeStart), maxTime)
+    } else {
+      updateScrollCanvas(visibleTimeStart, visibleTimeEnd)
+    }
+  }
+
+  moveResizeValidator = (action, item, time, resizeEdge) => {
+    if (time < new Date().getTime()) {
+      var newTime =
+        Math.ceil(new Date().getTime() / (15 * 60 * 1000)) * (15 * 60 * 1000)
+      return newTime
+    }
+
+    return time
+  }
+
+  itemRenderer = ({
+    item,
+    timelineContext,
+    itemContext,
+    getItemProps,
+    getResizeProps,
+  }) => {
+    const { left: leftResizeProps, right: rightResizeProps } = getResizeProps()
+    const backgroundColor = itemContext.selected ? itemContext.dragging ? 'red' : item.selectedBgColor : item.bgColor;
+    const borderColor = itemContext.resizing ? 'red' : item.color;
+    return (
+      <div
+        {...getItemProps({
+          style: {
+            backgroundColor,
+            color: item.color,
+            borderColor,
+            borderStyle: 'solid',
+            borderWidth: 1,
+            borderRadius: 4,
+            borderLeftWidth: itemContext.selected ? 3 : 1,
+            borderRightWidth: itemContext.selected ? 3 : 1,
+          }
+        }) }
+      >
+        {itemContext.useResizeHandle ? (
+          <div {...leftResizeProps} />
+        ) : null}
+
+        <div
+          style={{
+            height: itemContext.dimensions.height,
+            overflow: 'hidden',
+            paddingLeft:3,
+            textOverflow: 'ellipsis',
+            whiteSpace: 'nowrap'
+          }}
+        >
+          {itemContext.title}
+        </div>
+
+
+        {itemContext.useResizeHandle ? (
+          <div {...rightResizeProps} />
+        ) : null}
+      </div>
+    )
+  }
+
+  // groupRenderer = ({ group }) => {
+  //   return (
+  //     <div className='custom-group'>
+  //       {group.title}
+  //     </div>
+  //   )
+  // }
+
+  render() {
+    const { groups, items, defaultTimeStart, defaultTimeEnd } = this.state
+    console.log("render")
+    return (
+      <Timeline
+        groups={groups}
+        items={items}
+        keys={keys}
+        sidebarWidth={150}
+        sidebarContent={<div>Above The Left</div>}
+        // rightSidebarWidth={150}
+        // rightSidebarContent={<div>Above The Right</div>}
+
+        canMove
+        canResize="right"
+        canSelect
+        itemsSorted
+        itemTouchSendsClick={false}
+        stackItems
+        itemHeightRatio={0.75}
+        lineHeight={40}
+        showCursorLine
+        // resizeDetector={containerResizeDetector}
+
+        defaultTimeStart={defaultTimeStart}
+        defaultTimeEnd={defaultTimeEnd}
+        itemRenderer={this.itemRenderer}
+        // groupRenderer={this.groupRenderer}
+
+        onCanvasClick={this.handleCanvasClick}
+        onCanvasContextMenu={this.handleCanvasContextMenu}
+        onItemClick={this.handleItemClick}

```

Comment:
With this refactor, can we remove all `onItem*` props from the timeline component?  That would be a HUGE deal!

Reply:
will do

Reply:
I agree that this would be a huge deal but I don't think it is practical in the current implementation. The `onItem*` props are more of a timeline props with specific callback params to timeline. What I mean is that the `onItemClick` is not the same as the `onClick` dom event. I think the ideal scenario to achieve this would be having a new component that can register the render callback for item and accept all the `onItem*` props. What do you think?

Reply:
I agree. Maybe something to work towards but totally not needed as part of this PR.